[Sema] Warn about unexpected `.self` behavior with metatypes

davezarzycki opened the pull request:
When `Foo` is a type, programmers often type `Foo.Type.self` in an expression context when they mean `Foo.self`. Let's warn them.

Feedback would be appreciated about what a "good" warning message would look like.

Also, are there any good real-world reasons for programmers to work with meta-metatypes? How much effort should we go to let programmers suppress these two warnings?

davezarzycki commented:
As an aside, We could make the Swift compiler faster if we deprecated and eventually remove meta-metatypes. Doing so would allow us to:

1) OR a bit into a `TypeBase` pointer to represent the metatype
2) AND the pointer to get the instance type

The net result is less pointer chasing , less memory usage, and a simpler type system.

slavapestov commented:
> As an aside, we could make the Swift compiler faster if we deprecated and eventually remove meta-metatypes. 

This would prevent being able to bind a generic parameter to a metatype. For example, you couldn't have an array of metatypes.

davezarzycki commented:
> > As an aside, we could make the Swift compiler faster if we deprecated and eventually remove meta-metatypes.
> 
> This would prevent being able to bind a generic parameter to a metatype. For example, you couldn't have an array of metatypes.

Hi @slavapestov – Are you sure? I feel like I'm missing something. If I add the following assert on top of #32988 (and fix a few silly bugs that would have been caught by this pull request), then the following code doesn't trip over the assert:
```
diff --git i/lib/AST/ASTContext.cpp w/lib/AST/ASTContext.cpp
index bf3b609c68d30dd38eed3b1118be6a15ab9b2ab1..a4dc9aa319bc167508573f65658fd76366d6f590 100644
--- i/lib/AST/ASTContext.cpp
+++ w/lib/AST/ASTContext.cpp
@@ -2795,10 +2795,12 @@ AnyMetatypeType::AnyMetatypeType(TypeKind kind, const ASTContext *C,
 MetatypeType *MetatypeType::get(Type T, Optional<MetatypeRepresentation> Repr,
                                 const ASTContext &Ctx) {
   auto properties = T->getRecursiveProperties();
   auto arena = getArena(properties);

+  assert(!T->is<MetatypeType>());
+
   unsigned reprKey;
   if (Repr.hasValue())
     reprKey = static_cast<unsigned>(*Repr) + 1;
   else
     reprKey = 0;

```

The following test file type checks without crashing:

```
func foo<T>(_ value: [T]) {}

class Foo {}

foo([Foo.self, Foo.self])
```

davezarzycki commented:
If you'd like, we can move the aside about internal compiler performance to the forums. I'd like to focus on getting the warnings added :-)

davezarzycki commented:
Ping. Feedback would be appreciated on what these two diagnostics messages should be. I'd like them to essentially say "`XYZ.Type.self` produces a meta-metatype. This probably is not what you want" with a possible fix-it that removes `.Type`. What do you all think?

davezarzycki commented:
Ping?

davezarzycki commented:
@DougGregor @slavapestov @CodaFi 

Ping?

slavapestov commented:
@davezarzycki Imagine something like this compiled with optimizations:

```
func foo<T>(_: T) {
  print(T.self)
}

foo(Int.self)
```

The specializer will create a specialization of foo() for T := Int.Type; substituting that into the body will then produce `print(Int.Type.self)`, which will trip your assert.

davezarzycki commented:
Hi @slavapestov – I'm confused:

1) This pull request adds no asserts. So I'm not sure why you think your example code will "trip your assert". Perhaps you're referring to my earlier tangent in this PR about optimizing how the compiler represents metatypes?
2) The goal of this pull request is to only warning about explicit meta-metatype usage by the user. I wouldn't expect your example code to warn. And indeed, when I test your example with this PR, it does not warn.
3) I thought the language model was such that the output of your example code shouldn't depend on optimizations. What am I missing?
4) The output of your example (with top-of-tree) is `Int.Type` not `Int.Type.Type` which surprises me and probably you too!

davezarzycki commented:
I've updated the pull request after verifying that parenthesis silence the warning. I think the test suite fallout is reasonable, but reasonable people may disagree. Let's discuss:

@swift-ci please test

davezarzycki commented:
I've silenced the test suite failures with parenthesis so that people don't need to manually review the build failure to see the fallout. I've also fixed an apparent bug in a few of the stdlib unit tests.

@swift-ci please smoke test

davezarzycki commented:
Fixed an ASAN test (which I normally don't bother testing).

@swift-ci please smoke test

davezarzycki commented:
Second attempt at fixing the ASAN test. (I rebased the pull request and forgot to commit the speculative fix.)

@swift-ci please smoke test

davezarzycki commented:
Just tinkering with the code quality. @slavapestov – I think this is really ready now.

@swift-ci please smoke test

davezarzycki commented:
The linux failure looks spurious.

@swift-ci please smoke test linux

davezarzycki commented:
@slavapestov – I really think this is ready to go and it catches a bug pattern in the unit tests.

theblixguy commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -4098,6 +4098,14 @@ WARNING(class_super_not_usable_from_inline_warn,none,
         (bool))
 
+WARNING(meta_metatype,none,
+        "meta-metatype is probably not what you want", ())
```
IMO the wording of this diagnostic doesn’t quite fit the style of the existing diagnostics we emit in unexpected situations. I think it should be something along the lines of “type of {thing} is a meta-metatype, which may be unexpected”.

davezarzycki replied:
Sure. That kind of feedback is what I initially requested and wanted when I opened this pull request. Thanks.

theblixguy commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -4098,6 +4098,14 @@ WARNING(class_super_not_usable_from_inline_warn,none,
+WARNING(metatype_self,none,
+        "self of a metatype produces a meta-metatype", ())
+NOTE(metatype_self_silence,none,
+     "add parentheses to silence this warning", ())
```
There needs to be tests for all of these diagnostics and the associated fix-its.

Also, it would be great if you copied these diagnostics over to the diagnostic localization file (located at `localization/diagnostics/en.yaml`) as well.

davezarzycki replied:
Okay. Fixed but:

1) There really ought to be a test that ensures that the "localized" English content doesn't drift.
2) Better still, why isn't the English "localization" auto generated?

hborla replied:
`en.yaml` isn't auto generated because we intend to remove the message strings from `DiagnosticsSema.def`. The goal is for `DiagnosticsSema.def` to only contain the diagnostic ID and argument types, and the `.yaml` files will contain the message strings so that we don't need to rebuild the compiler every time we want to tweak the wording of a message. It's in this inconvenient state right now because the transition is still a work in progress

davezarzycki replied:
Interesting. I still wish we'd auto-generate the YAML until we're ready to flip the switch. The duplication is just asking for bugs. And yes, I am assuming that the YAML file WAS auto-generated at one point and we could have saved that script, but perhaps I'm assuming too much.

theblixguy commented on lib/Sema/CSGen.cpp:
```diff
@@ -1712,6 +1712,38 @@ namespace {
+        .fixItInsert(typeExpr->getStartLoc(), "(")
+        .fixItInsertAfter(typeExpr->getEndLoc(), ")");
+
+      return result;
```
What if the user provides an explicit type annotation? For example:

```swift
let foo: AnyObject.Type.Protocol = AnyObject.Type.self
```

Should the warning still be emitted?

davezarzycki replied:
I think so. Meta-metatypes (and meta-meta-metatypes etc, etc) are almost completely devoid of utility in the user-facing language and likely to remain that way. If the user tries to use them, they're probably wandering into the weeds and a warning is appropriate, even if explicit types are used.

davezarzycki replied:
If I might get opinionated, I'd deprecate meta-metatypes (and meta-meta-metatypes, etc, etc) in the user-facing language. They don't have any members, nor can they be defined or extended. They also don't really serve a purpose other than to expose how the compiler models metatypes internally.

davezarzycki committed:
[Sema] Warn about unexpected `.self` behavior with metatypes

Users are sometimes confused by the fact that types are implicitly
metatypes in expression contexts.

davezarzycki commented:
I've simplified this pull request to focus on unexpected `.self` behavior for users and I believe that I've addressed all feedback to date.

I removed the warning about TypeReprs that are explicit meta-metatypes because I couldn't think of a "good" warning that isn't about deprecation and that kind of change needs to go through formal review. I also think that explicit meta-metatypes in type contexts are more rare than the accidental meta-metatypes via `.self`.

@swift-ci please smoke test

davezarzycki commented:
@slavapestov @xedin @DougGregor @CodaFi – Ping.

I've narrowed this pull request and I think I've addressed all of the feedback to date. Can somebody please okay this for merging? Thanks!

davezarzycki commented:
Ping? This is a really simple change and it exposes a real bug in the stdlib unit tests.

davezarzycki commented:
Ping

xedin reviewed (commented):
Sorry for the delay, I have left a couple of suggestions inline. I have mixed feelings about this warning because in `warnings-as-errors` mode it is effectively a source compatibility regression even though the code might use `.Type.self` intentionally, from that point of view I'm not sure this is really worth it...

xedin commented:
I'm indeed skeptical about the trade-off here but I don't really want this PR to be closed based on my opinion alone. If anybody disagrees (@DougGregor, @slavapestov, @hborla for example) I'd be fine merging this after changes I have mentioned are implemented...

davezarzycki commented:
Hi @DougGregor, @slavapestov, and @hborla –

Ping. @xedin asked for your opinions about whether this warning is worthwhile. I'd appreciate your input. Thanks!

davezarzycki commented:
Hi @DougGregor, @slavapestov, and @hborla –

Ping. @xedin asked for your opinions about whether this warning is worthwhile. I'd appreciate your input. Thanks!

hborla commented:
I apologize for the delayed response!

Hmm, I think I'm in the same boat as Pavel and Robert. It feels weird to me to warn about a valid feature in the language, especially when the user wrote an explicit type annotation. If it's truly the case that meta-metatypes are totally useless (which, while I understand the motivation for this PR, I'm not yet 100% convinced of), then I think we should just ban them outright instead of warning.

rjmccall commented:
As Slava observes, we can't actually eliminate meta-metatypes from the language without making the metatype system incoherent if you happen to be in generic code over metatypes.

Maybe the diagnostic should be a special-case error for type mismatches that differ by a metatype level when the source operand is a `Type.self` expression?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
